keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,874 @@
+types. We will modify this `GroupVersionForDiscovery` type to include
+a list of
+[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
+This list of `APIResource` is what is currently published at the
+`/apis/<group>/version` endpoint and is what will be aggregated into
+the new endpoint that we publish.
+
+The endpoint will also publish an ETag calculated based on a hash of
+the data for clients.
+
+This is approximately what the new API will look like (conflicting names will be renamed)
+
+```go
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+type APIGroupList struct {
+	TypeMeta `json:",inline"`
+	// groups is a list of APIGroup.
+	Groups []APIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+}
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type APIGroup struct {
+	TypeMeta `json:",inline"`
+	// name is the name of the group.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// versions are the versions supported in this group.
+	// This will be sorted in descending order based on the preferred version
+	Versions []GroupVersionForDiscovery `json:"versions" protobuf:"bytes,2,rep,name=versions"`
+	// PreferredVersion will be removed for the new Discovery API
+	// The GroupVersionForDiscovery will be sorted based on the preferred version
+	// PreferredVersion GroupVersionForDiscovery `json:"preferredVersion,omitempty" protobuf:"bytes,3,opt,name=preferredVersion"`
+
+ 	// ServerAddresssByClientCIDRs will be removed for the new Discovery API
+	//ServerAddressByClientCIDRs []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs,omitempty" protobuf:"bytes,4,rep,name=serverAddressByClientCIDRs"`
+}
+
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensibility.
+type GroupVersionForDiscovery struct {
+	// groupVersion specifies the API group and version in the form "group/version"
+	// This will be removed for the new discovery
+	// GroupVersion string `json:"groupVersion" protobuf:"bytes,1,opt,name=groupVersion"`
+	// version specifies the version in the form of "version". This is to save
+	// the clients the trouble of splitting the GroupVersion.
+	Version string `json:"version" protobuf:"bytes,2,opt,name=version"`
+	// resources contains the name of the resources and if they are namespaced.
+	APIResources []APIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`
+
+ 	// LastContacted is the last time that the apiserver has successfully reached the
+ 	// corresponding group version's discovery document. This will be nil if the group-version
+ 	// has not been aggregated yet. To maintain consistency across scenarios with multiple
```

Comment:
If it is nil, does that imply that APIResources above is an empty list?

Reply:
Yes, added a note